action.yml

```diff
@@ -0,0 +1,199 @@
+name: Merge up
+description: GitHub Action to merge-up supported branches in a repository
+
+runs:
+  using: composite
+  steps:
+
+    - name: Install PHP
+      uses: shivammathur/setup-php@1a18b2267f80291a81ca1d33e7c851fe09e7dfc4 # v2.22.0
+      with:
+        php-version: '8.1'
+
+    - name: Determine if should merge-up
+      id: determine
+      shell: bash
+      env:
+        GITHUB_REPOSITORY: ${{ github.repository }}
+        GITHUB_REF_NAME: ${{ github.ref_name }}
+      run: |
+        # The minimum cms major with commercial support - configured at a global level
+        # Change this when major version support changes
+        MINIMUM_CMS_MAJOR=4
+
+        # Get the default branch from GitHub API
+        # We need to make an API call rather than just assume that the current branch is the default
+        # because this workflow may be triggered by workflow_dispatch on any branch
+        RESP_CODE=$(curl -w %{http_code} -s -o __base.json \
+        -X GET "https://api.github.com/repos/$GITHUB_REPOSITORY" \
+        -H "Accept: application/vnd.github+json" \
+        -H "Authorization: Bearer ${{ github.token }}" \
+        -H "X-GitHub-Api-Version: 2022-11-28" \
+        )
+        if [[ $RESP_CODE != "200" ]]; then
+          echo "Unable to read list of tags - HTTP response code was $RESP_CODE"
+          exit 1
+        fi
+
+        # Gets all tags from GitHub API
+        # https://docs.github.com/en/rest/git/tags?apiVersion=2022-11-28
+        RESP_CODE=$(curl -w %{http_code} -s -o __tags.json \
+        -X GET "https://api.github.com/repos/$GITHUB_REPOSITORY/tags?per_page=100" \
+        -H "Accept: application/vnd.github+json" \
+        -H "Authorization: Bearer ${{ github.token }}" \
+        -H "X-GitHub-Api-Version: 2022-11-28" \
+        )
+        if [[ $RESP_CODE != "200" ]]; then
+          echo "Unable to read list of tags - HTTP response code was $RESP_CODE"
+          exit 1
+        fi
+
+        # Gets all branches from GitHub API
+        # https://docs.github.com/en/rest/branches/branches?apiVersion=2022-11-28#list-branches
+        RESP_CODE=$(curl -w %{http_code} -s -o __branches.json \
+        -X GET "https://api.github.com/repos/$GITHUB_REPOSITORY/branches?per_page=100" \
+        -H "Accept: application/vnd.github+json" \
+        -H "Authorization: Bearer ${{ github.token }}" \
+        -H "X-GitHub-Api-Version: 2022-11-28" \
+        )
+        if [[ $RESP_CODE != "200" ]]; then
+          echo "Unable to read list of tags - HTTP response code was $RESP_CODE"
+          exit 1
+        fi
+
+        DEFAULT_BRANCH=$(jq -r .default_branch __base.json)
+        echo "DEFAULT_BRANCH is $DEFAULT_BRANCH"
+        rm __base.json
+
+        if [[ $DEFAULT_BRANCH != $GITHUB_REF_NAME ]]; then
+          echo "Current branch $GITHUB_REF_NAME is not the same as default branch $DEFAULT_BRANCH"
+          exit 1
```

Comment:
Assumption here is wrong

This workflow is only restred to _run_ on the default branch - however later on it will still do a full history `git clone` (via `actions/checkout`) and then manually `git checkout` the branches that need to be merged-up

Being on the default branch just protects against potentially weird stuff happening later, such as manually running this job in the future on an older branch (e.g. on CMS 5 when the default has moved on to CMS 6) - while the older branch may be say stuck on `v1` on this action while the newer branch uses `v2`

Reply:
> Default branch is also used in funcs.php to work out the minimum cms major to merge up from e.g. 4 - once we change the default branch to 6 then the default supported major becomes 5

This is the part I'm talking about, really. I'm fine with having a mechanism to restrict to only run this on the default branch - but we can't safely assume that branch is _the most up to date branch_, and therefore we can't use it to identify the minimum CMS major to merge up from.

Reply:
I'm not sure I follow

My thinking here is that this is the current state of play - it'll merge up from 4 to 5:
- CMS 5 << default branch
- CMS 4 << minimum supported

Once we create CMS 6 branches it will be this, it'll merge up from 4 to 5 to 6
- CMS 6
- CMS 5 << default branch
- CMS 4 << minimum supported

When CMS 6 is goes stable,  but before we've switched over all the default branches it will be this, and it'll merge up from 4 to 5 to 6 (more than we technically need to since 4 will be out of support at this point)
- CMS 6
- CMS 5 << default branch  << minimum supported
- CMS 4

When we switch over the default branches to CMS 6 it will be this, and it'll merge up from 5 to 6
- CMS 6 << default branch
- CMS 5 << minimum supported

Reply:
I don't think that's really an issue?

Take for instance symbiote/advanced workflow where the default branch `5` is CMS 4 instead of what it should be `6` for CMS 5

Since action.yml defines `MINIMUM_CMS_MAJOR=4` then it won't merge-up anything from CMS 3.  It'll only start from CMS 4.

I've just added a unit test that maps the the currently state for symbiote/silverstripe-advanced-workflow called `Module where default branch has not been changed from CMS 4 and there is a new CMS 6 branch` to prove this work as they should

Reply:
Ahh, yup good point. I'd forgotten about that hardcoded value.
Thanks for adding the test as an additional safety net.